File: leetcode/problems/4-median-of-two-sorted-arrays/solution.c
#include <stdio.h>

#define MAX 2000

void print_array(int nums[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", nums[i]);
    }
    printf("\n");
}

double find_median_sorted_array(int *nums1, int n1, int *nums2, int n2) {
    int nums[MAX];

    int i = 0, i1 = 0, i2 = 0;
    while (i1 < n1 || i2 < n2) {
        if (i1 < n1 && i2 < n2) {
            nums[i++] = nums1[i1] < nums2[i2] ? nums1[i1++] : nums2[i2++];
        } else if (i1 < n1) {
            nums[i++] = nums1[i1++];
        } else if (i2 < n2) {
            nums[i++] = nums2[i2++];
        }
    }

    if (i % 2 != 0) {
        return (double) nums[(i - 1) / 2];
    }
    
    return ((double) nums[(i - 1) / 2] + (double) nums[(i - 1) / 2 + 1]) / (double) 2;
}

int main() {

    int T, tc = 1;

    if (scanf("%d", &T) != 1) return 1;

    while (T--) {

        int nums1[MAX / 2];
        int n1 = 0;
        char c;
        while (scanf("%d%c", &nums1[n1++], &c) == 2) {
            if (c == '\n') break;
        }

        int nums2[MAX / 2];
        int n2 = 0;
        while (scanf("%d%c", &nums2[n2++], &c) == 2) {
            if (c == '\n') break;
        }

        double m = find_median_sorted_array(nums1, n1, nums2, n2);
        printf("Case #%d:\n", tc++);
        printf("%.5f\n", m);

    }

    return 0;
}

File: leetcode/problems/4-median-of-two-sorted-arrays/solution.go
package main

import (
	"bufio"
	"os"
	"fmt"
	"log"
	"strconv"
	"strings"
)

var (
	in = bufio.NewReader(os.Stdin)
)

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	nums := make([]int, len(nums1) + len(nums2))

	n1 := len(nums1)
	n2 := len(nums2)
	i := 0
	i1 := 0
	i2 := 0
	for ; i1 < n1 || i2 < n2; {
		if i1 < n1 && i2 < n2 {
			if nums1[i1] < nums2[i2] {
				nums[i] = nums1[i1]
				i++
				i1++
			} else {
				nums[i] = nums2[i2]
				i++
				i2++
			}
		} else if i1 < n1 {
			nums[i] = nums1[i1]
			i++
			i1++
		} else if i2 < n2 {
			nums[i] = nums2[i2]
			i++
			i2++
		}
	}

	if i % 2 != 0 {
		return float64(nums[(i - 1) / 2])
	}
	return float64((nums[(i - 1) / 2] + nums[(i - 1) / 2 + 1])) / float64(2);
}

func main() {
	T := readInt(in)

	for tc := 1; tc <= T; tc++ {

		nums1 := readLineInt(in)
		nums2 := readLineInt(in)

		m := findMedianSortedArrays(nums1, nums2)
		fmt.Printf("Case #%d:\n", tc)
		fmt.Printf("%.5f\n", m)
	}
}

func readLine(in *bufio.Reader) []string {
	line, _ := in.ReadString('\n')
	line = strings.ReplaceAll(line, "\r", "")
	line = strings.ReplaceAll(line, "\n", "")

	datas := strings.Split(line, " ")
	return datas
}

func readInt(in *bufio.Reader) int {
	num := readLine(in)
	if num[0] == "" {
		log.Fatal("no data")
	}
	n, _ := strconv.Atoi(num[0])
	return n
}

func readLineInt(in *bufio.Reader) []int {
	nums := readLine(in)
	if nums[0] == "" {
		return make([]int, 0)
	}
	arr := make([]int, len(nums))

	for i, n := range nums {
		val, _ := strconv.Atoi(n)
		arr[i] = val
	}

	return arr;
}
